presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TranslateExpressions.java

```diff
@@ -271,6 +292,213 @@ public Result apply(ValuesNode valuesNode, Captures captures, Context context)
+            verify(lambdaExpressions.size() == lambdaInterfaces.size());
+            ImmutableMap.Builder<NodeRef<Expression>, Type> buildr = ImmutableMap.<NodeRef<Expression>, Type>builder();
+
+            for (int i = 0; i < lambdaExpressions.size(); i++) {
+                LambdaExpression lambdaExpression = lambdaExpressions.get(i);
+                FunctionType functionType = functionTypes.get(i);
+
+                // To compile lambda, LambdaDefinitionExpression needs to be generated from LambdaExpression,
+                // which requires the types of all sub-expressions.
+                //
+                // In project and filter expression compilation, ExpressionAnalyzer.getExpressionTypesFromInput
+                // is used to generate the types of all sub-expressions. (see visitScanFilterAndProject and visitFilter)
+                //
+                // This does not work here since the function call representation in final aggregation node
+                // is currently a hack: it takes intermediate type as input, and may not be a valid
+                // function call in Presto.
+                //
+                // TODO: Once the final aggregation function call representation is fixed,
+                // the same mechanism in project and filter expression should be used here.
+                verify(lambdaExpression.getArguments().size() == functionType.getArgumentTypes().size());
+                Map<NodeRef<Expression>, Type> lambdaArgumentExpressionTypes = new HashMap<>();
+                Map<Symbol, Type> lambdaArgumentSymbolTypes = new HashMap<>();
+                for (int j = 0; j < lambdaExpression.getArguments().size(); j++) {
+                    LambdaArgumentDeclaration argument = lambdaExpression.getArguments().get(j);
+                    Type type = functionType.getArgumentTypes().get(j);
+                    lambdaArgumentExpressionTypes.put(NodeRef.of(argument), type);
+                    lambdaArgumentSymbolTypes.put(new Symbol(argument.getName().getValue()), type);
+                }
+                // the lambda expression itself
+                buildr.put(NodeRef.of(lambdaExpression), functionType)
+                        // expressions from lambda arguments
+                        .putAll(lambdaArgumentExpressionTypes)
+                        // expressions from lambda body
+                        .putAll(getExpressionTypes(
+                                context.getSession(),
+                                metadata,
+                                sqlParser,
+                                TypeProvider.copyOf(lambdaArgumentSymbolTypes),
+                                lambdaExpression.getBody(),
+                                emptyList(),
+                                NOOP));
+            }
+            return buildr.build();
+        }
+        return ImmutableMap.of();
+    }
+
     private RowExpression toRowExpression(Expression expression, Rule.Context context)
```

Comment:
This function body should be delegated to the one below by having lambdaTypes is an empty map